### PR TITLE
Remove metastore transactions whenever possible

### PR DESCRIPTION
**Pull request by guilload**
### Description
Remove metastore transactions whenever possible. I will remove the transactions for `mark_splits_for_deletion` and `delete_splits` in follow-up PRs.

### How was this PR tested?
Ran Postgres metastore tests.

### CODE REVIEW

File: quickwit/quickwit-metastore/src/metastore/mod.rs [hunk 1]
```diff
@@ -279,7 +279,7 @@ pub trait Metastore: Send + Sync + 'static {
 /// A query builder for listing splits within the metastore.
 pub struct ListSplitsQuery<'a> {
     /// The index to get splits from.
-    pub index: &'a str,
+    pub index_id: &'a str,
```
Comment (ChillFish8) [inline quickwit/quickwit-metastore/src/metastore/mod.rs:282]: This was annoying me a bit as well 😅 Although I'm not a huge fan of adding this small unrelated refactor in with the rest of the changes.

File: quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs [hunk 2]
```diff
@@ -564,59 +539,49 @@ impl Metastore for PostgresqlMetastore {
...
+            .fetch_all(&self.connection_pool)
+            .await?;
+        pg_indexes
+            .into_iter()
+            .map(|pg_index| pg_index.index_metadata())
+            .collect()
     }
 
-    #[instrument(skip(self),fields(index_id=index_metadata.index_id.as_str()))]
+    #[instrument(skip(self),fields(index_id=index_metadata.index_id))]
```
Comment (ChillFish8) [inline quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:553]: If we're adding some formatting/whitespace to the instrument macros while we go, we should probably add them to any of the other instruments we add here.

```suggestion
    #[instrument(skip(self), fields(index_id=index_metadata.index_id))]
```

File: quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs [hunk 3]
```diff
@@ -564,59 +539,49 @@ impl Metastore for PostgresqlMetastore {
...
+        if delete_res.rows_affected() == 0 {
+            return Err(MetastoreError::IndexDoesNotExist {
+                index_id: index_id.to_string(),
+            });
+        }
+        Ok(())
     }
 
-    #[instrument(skip(self, split_metadata),fields(split_id=split_metadata.split_id))]
+    #[instrument(skip(self, split_metadata),fields(index_id=index_id, split_id=split_metadata.split_id))]
```
Comment (ChillFish8) [inline quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:584]: formatting/whitespace

File: quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs [hunk 4]
```diff
@@ -872,65 +859,58 @@ impl Metastore for PostgresqlMetastore {
...
-                .fetch_one(tx)
-                .await
-                .map(|row| (row.get(0), row.get(1)))
-                .map_err(|err| convert_sqlx_err(&delete_query.index_id, err))?;
-            Ok(DeleteTask {
-                create_timestamp: create_timestamp.assume_utc().unix_timestamp(),
-                opstamp: opstamp as u64,
-                delete_query: Some(delete_query),
-            })
+        let (create_timestamp, opstamp): (sqlx::types::time::PrimitiveDateTime, i64) = sqlx::query(
```
Comment (ChillFish8) [inline quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:894]: I think we can use `query_as` here and remove the resulting `map()` method where we're manually getting the rows.

File: quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs [hunk 5]
```diff
@@ -872,65 +859,58 @@ impl Metastore for PostgresqlMetastore {
...
+            "INSERT INTO delete_tasks (index_id, delete_query_json) VALUES ($1, $2) RETURNING \
+             create_timestamp, opstamp",
```
Comment (ChillFish8) [inline quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:896]: Nitpick: The formatting of the SQL here done by rustfmt hasn't done a lot of good for reading.

```suggestion
            "INSERT INTO delete_tasks (index_id, delete_query_json) VALUES ($1, $2) \
             RETURNING create_timestamp, opstamp",
```

File: quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs [hunk 6]
```diff
@@ -872,65 +859,58 @@ impl Metastore for PostgresqlMetastore {
...
+        )
+        .bind(&delete_query.index_id)
+        .bind(&delete_query_json)
+        .fetch_one(&self.connection_pool)
+        .await
+        .map(|row| (row.get(0), row.get(1)))
```
Comment (ChillFish8) [inline quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:902]: This can be removed if we use `query_as`.

File: quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs [hunk 7]
```diff
@@ -940,82 +920,78 @@ impl Metastore for PostgresqlMetastore {
...
+        pg_delete_tasks
+            .into_iter()
+            .map(|pg_delete_task| pg_delete_task.try_into())
+            .collect::<MetastoreResult<_>>()
     }
 
     /// Returns `num_splits` published splits with `split.delete_opstamp` < `delete_opstamp`
-    /// ordered by ASC `split.delete_opstamp`.
-    #[instrument(skip(self))]
+    /// ordered by ascending `split.delete_opstamp` and `split.publish_timestamp`.
```
Comment (ChillFish8) [inline quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:984]: Nitpick: I don't think the wording reads great here, it reads like it's part of the first sentence but probably should be read as a separate sentence, I think something along these lines probably makes it a bit better, but it's not the end of the world.

```suggestion
    /// Returns `num_splits` published splits with `split.delete_opstamp` < `delete_opstamp`.
    ///
    /// Results are ordered by ascending `split.delete_opstamp` and `split.publish_timestamp` values.
```

or

```suggestion
    /// in ascending order of `split.delete_opstamp` and `split.publish_timestamp`.
```

File: quickwit/quickwit-metastore/src/metastore/mod.rs (hunk 1 above)
Comment (guilload) [reply quickwit/quickwit-metastore/src/metastore/mod.rs:282]: Yes, I can split the PRs. This is already in an isolated commit.

File: quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs (hunk 6 above)
Comment (guilload) [reply quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:902]: I would need to implement `FromRow` for the tuple. Same for u64.

Comment (ChillFish8) [reply quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:902]: SQLx provides a blanket implementation on [tuples](https://docs.rs/sqlx/latest/sqlx/trait.FromRow.html#impl-FromRow%3C%27r%2C%20R%3E-for-(T1%2C%20T2)) so I think this should 'just work' for the DateTime and i64.

Comment (guilload) [reply quickwit/quickwit-metastore/src/metastore/postgresql_metastore.rs:904]: Indeed. Thanks. TIL, there's also `query_scalar` for a single value per row.

**Review by ChillFish8 — APPROVED** [review]
